### PR TITLE
drivers: add driver for Synopsys DWC2 that is used as USB OTG peripheral on STM32 and ESP32x SoCs

### PR DESCRIPTION
**Pull request by gschorcht**
### Contribution description

This PR adds a driver for the Synopsys DWC2 IP core. It is an alternative to PR #18624 which avoids a lot of code duplication. STM32 as well as ESP32x SoCs use the Synopsys DWC2 core. According to [tinyUSB documentation](https://github.com/hathach/tinyusb/blob/master/docs/reference/supported.rst), EFM32GG12 and GD32VF103 also integrate the Synopsys DWC2 core. 

Instead of copying the code from STM32's `periph/usbdev_otg.c` and modifying it very slightly only for ESP32x SoCs (as done in PR #18624), the code is now moved to a common location and extended by only a very small number of MCU-Specific changes. The Namespace is changed from `stm32` to `dwc2`. STM32 and ESP32x SoCs use the driver as `periph_usbdev` driver.

### Testing procedure

The following tests have to work for an ESP32 node as well as a STM32 node to check whether it is still working for STM32.
1. Flash `tests/usbus_hid`
    ```
    BOARD=esp32s3-devkit make -C tests/usbus_hid flash term
    ```
    ```
    BOARD=nucleo-f767zi make -C tests/usbus_hid flash term
    ```
    Use the `dmesg` command and check the USB HID devices was found
    ```python
    [1230033.363705] usb 1-2.4: USB disconnect, device number 91
    [1230033.625919] usb 1-2.4: new full-speed USB device number 92 using xhci_hcd
    [1230033.732624] usb 1-2.4: New USB device found, idVendor=1209, idProduct=7d01, bcdDevice= 1.00
    [1230033.732627] usb 1-2.4: New USB device strings: Mfr=3, Product=2, SerialNumber=4
    [1230033.732629] usb 1-2.4: Product: esp32s3-devkit
    [1230033.732630] usb 1-2.4: Manufacturer: RIOT-os.org
    [1230033.732631] usb 1-2.4: SerialNumber: A5894EECB420ECF1
    [1230033.739837] hid-generic 0003:1209:7D01.00D9: hiddev2,hidraw5: USB HID v1.10 Device [RIOT-os.org esp32s3-devkit] on usb-0000:00:14.0-2.4/input0
    ```
    and that there is a new `/dev/hidrawX` device to which you can write as `root` characters that are echoed in console window, for example:
    ```
    # echo -n "Hello" > /dev/hidraw5
    ```
    ```
    main(): This is RIOT! (Version: 2022.10-devel-619-gca66ff-cpu/esp32/periph_usbdev)
    RIOT USB HID echo test
    Write input to the hidraw device under /dev/hidrawX, and see if it gets echoed here
    USB_HID rx_cb: Test argument 
    Msg received via USB HID: 
    Hello
    ```
2. Flash `tests/usbus_ecm`
    ```
    BOARD=esp32s3-devkit make -C tests/usbus_hid flash term
    ```
    ```
    BOARD=nucleo-f767zi make -C tests/usbus_hid flash term
    ```
   On the host computer, using tools such as `ethtool` must show the USB CDC ECM interface as link detected:
   ```shell
   # ethtool enp0s20u9u4
   Settings for enp0s20u9u4:
           Current message level: 0x00000007 (7)
                                  drv probe link
           Link detected: yes
   ```
   You should be able to ping the ESP32x node from the host and vice versa using the IPv6 address of the USB CDC ECM interface.

### CODE REVIEW

File: cpu/esp32/Makefile.features [hunk 1]
```diff
@@ -41,6 +41,10 @@ else ifneq (,$(filter esp32c3 esp32s3,$(CPU_FAM)))
   FEATURES_PROVIDED += esp_ble_esp32c3
 endif
 
+ifneq (,$(filter esp32s2 esp32s3,$(CPU_FAM)))
+  FEATURES_PROVIDED += periph_usbdev
```
Comment (benpicco) [inline cpu/esp32/Makefile.features:45]: We still have to provide this at the board level - not every board has a target USB connector (e.g. my esp32-s2-wroom based [lilygo board](https://aliexpress.com/item/1005004246784287.html) does not)

File: boards/common/esp32s2/include/periph_conf_common.h [hunk 2]
```diff
@@ -337,6 +337,81 @@ static const uart_conf_t uart_config[] = {
...
+ * @brief   Total size of the FIFO
+ */
+#ifndef DWC2_USB_OTG_FS_TOTAL_FIFO_SIZE
+#define DWC2_USB_OTG_FS_TOTAL_FIFO_SIZE (1024U)
+#endif
+
+/**
+ * @brief   Buffers have to be word aligned for DMA
+ */
+#define USBDEV_CPU_DMA_ALIGNMENT        (4)
```
Comment (benpicco) [inline boards/common/esp32s2/include/periph_conf_common.h:398]: This is not board specific, it should go to `periph_cpu_esp32s2.h`

**Review by benpicco — COMMENTED** [review]
Awesome, that's the way to go :smiley: 

Works like a charm (I only tested on STM32, the only ESP32x board I have that has a USB port directly connected to the MCU is ESP32-C3), just two small nitpicks:

File: drivers/include/usbdev_synopsys_dwc2.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup drivers_periph_usbdev
+ * @{
+ *
+ * @file
+ * @brief   Low level USB FS/HS driver definitions for MCUs with Synopsys DWC2 IP core
```
Comment (benpicco) [inline drivers/include/usbdev_synopsys_dwc2.h:15]: Synopsys needs to be added to `dist/tools/codespell/ignored_words.txt`

**Comment by benpicco** [comment]
Please squash!

File: boards/common/esp32s3/include/periph_conf_common.h [hunk 4]
```diff
@@ -337,6 +337,84 @@ static const uart_conf_t uart_config[] = {
...
+ * @brief   Total size of the FIFO
+ */
+#ifndef DWC2_USB_OTG_FS_TOTAL_FIFO_SIZE
+#define DWC2_USB_OTG_FS_TOTAL_FIFO_SIZE (1024U)
+#endif
+
+/**
+ * @brief   Buffers have to be word aligned for DMA
+ */
+#define USBDEV_CPU_DMA_ALIGNMENT        (4)
```
Comment (benpicco) [inline boards/common/esp32s3/include/periph_conf_common.h:399]: Those are still CPU properties, not board properties

**Review by benpicco — APPROVED** [review]
Please squash!

**Comment by gschorcht** [comment]
Oops ...
```
-- running on worker comsys thread 7, build number 395.
make: Entering directory '/tmp/dwq.0.9713961674827482/55744c144780056868086f3022f1d3e8/examples/rust-gcoap'
make: Leaving directory '/tmp/dwq.0.9713961674827482/55744c144780056868086f3022f1d3e8/examples/rust-gcoap'
make: Entering directory '/tmp/dwq.0.9713961674827482/55744c144780056868086f3022f1d3e8/examples/rust-gcoap'
Building application "rust_gcoap" for "native" with MCU "native".

ccache: error: Failed to create temporary file for /cache/.ccache/tmp/tmp.cpp_stdout.8zgYzC: No space left on device
```

**Comment by benpicco** [comment]
Yea comsys ran out of space during the update to Ubuntu 22.04

**Comment by maribu** [comment]
Looks like the auto-merge of the jammy fixes came in to late. I'll restart the CI run, so that no unrelated issued pop up in the build log.

**Comment by benpicco** [comment]
Looks like CI is back to normal now - thank you for providing all this 😃

Btw: Does the ESP-C3 use the same peripheral or does this require more changes?

**Comment by gschorcht** [comment]
> Btw: Does the ESP-C3 use the same peripheral or does this require more changes?

Unfortunatly not. ESP32-C3 uses an special USB Serial/JTAG controller which is an USB CDC ACM device with fixed configuration. If at all, we could only develop a driver which then only supports the CDC ACM class. ESP32-S2 and ESP32-S3 have this USB Serial/JTAG controller in addition to the USB OTG controller.

File: boards/esp32s2-devkit/Makefile.features [hunk 5]
```diff
@@ -40,3 +40,7 @@ FEATURES_PROVIDED += periph_pwm
 FEATURES_PROVIDED += periph_spi
 
 FEATURES_PROVIDED += arduino
+
+ifneq (,$(filter esp32s2-devkitc-%,$(BOARD_VERSION)))
+  FEATURES_PROVIDED += tinyusb_device
```
Comment (benpicco) [inline boards/esp32s2-devkit/Makefile.features:45]: Now that's a different PR :wink:

Comment (gschorcht) [reply boards/esp32s2-devkit/Makefile.features]: Oops. I squashed this very small fix directly.

**Comment by gschorcht** [comment]
@benpicco Thanks for reviewing and merging.

**Pull request merged**